src/target/riscv/riscv.c

```diff
@@ -4376,96 +4388,160 @@ static bool gdb_regno_cacheable(enum gdb_regno regno, bool is_write)
+ */
 int riscv_get_register(struct target *target, riscv_reg_t *value,
 		enum gdb_regno regid)
 {
 	RISCV_INFO(r);
+	assert(r->get_register);
 
 	keep_alive();
 
-	struct reg *reg = &target->reg_cache->reg_list[regid];
+	if (!target->reg_cache) {
+		assert(!target_was_examined(target));
+		LOG_TARGET_DEBUG(target, "No cache, reading %s from target",
+				gdb_regno_name(regid));
+		return r->get_register(target, value, regid);
+	}
+
+	struct reg *reg = get_reg_cache_entry(target, regid);
 	if (!reg->exist) {
-		LOG_DEBUG("[%s] %s does not exist.",
-				  target_name(target), gdb_regno_name(regid));
+		LOG_TARGET_DEBUG(target, "Register %s does not exist.", reg->name);
 		return ERROR_FAIL;
 	}
 
-	if (reg && reg->valid) {
+	if (reg->valid) {
 		*value = buf_get_u64(reg->value, 0, reg->size);
-		LOG_DEBUG("[%s] %s: %" PRIx64 " (cached)", target_name(target),
-				  gdb_regno_name(regid), *value);
+		LOG_TARGET_DEBUG(target, "Read %s: 0x%" PRIx64 " (cached)", reg->name,
+				*value);
 		return ERROR_OK;
 	}
 
-	/* TODO: Hack to deal with gdb that thinks these registers still exist. */
-	if (regid > GDB_REGNO_XPR15 && regid <= GDB_REGNO_XPR31 &&
-			riscv_supports_extension(target, 'E')) {
-		*value = 0;
-		return ERROR_OK;
-	}
+	LOG_TARGET_DEBUG(target, "Reading %s from target", reg->name);
+	if (r->get_register(target, value, regid) != ERROR_OK)
+		return ERROR_FAIL;
 
-	int result = r->get_register(target, value, regid);
+	buf_set_u64(reg->value, 0, reg->size, *value);
+	reg->valid = gdb_regno_cacheable(regid, /* is write? */ false);
+	reg->dirty = false;
 
-	if (result == ERROR_OK) {
-		/* Update the cache in case we're called from
-		 * riscv_save_register(). */
-		buf_set_u64(reg->value, 0, reg->size, *value);
-		reg->valid = gdb_regno_cacheable(regid, false);
-	}
-
-	LOG_DEBUG("[%s] %s: %" PRIx64, target_name(target),
-			gdb_regno_name(regid), *value);
-	return result;
+	LOG_TARGET_DEBUG(target, "Read %s: 0x%" PRIx64, reg->name, *value);
+	return ERROR_OK;
 }
 
+/**
+ * This function is used to save the value of a register in cache. The register
+ * is marked as dirty, and writeback is delayed for as long as possible.
+ */
 int riscv_save_register(struct target *target, enum gdb_regno regid)
 {
 	RISCV_INFO(r);
 	riscv_reg_t value;
 	if (!target->reg_cache) {
 		assert(!target_was_examined(target));
+		/* To create register cache it is needed to examine the target first,
+		 * therefore during examine, any needed register should be saved
+		 * manually
+		 */
 		return ERROR_OK;
 	}
 
-	struct reg *reg = &target->reg_cache->reg_list[regid];
-	LOG_DEBUG("[%s] save %s", target_name(target), reg->name);
+	struct reg *reg = get_reg_cache_entry(target, regid);
+
+	LOG_TARGET_DEBUG(target, "Saving %s", reg->name);
 	if (riscv_get_register(target, &value, regid) != ERROR_OK)
 		return ERROR_FAIL;
 
-	if (!reg->valid)
+	if (!reg->valid) {
+		LOG_TARGET_DEBUG(target, "%s can't be cached", reg->name);
 		return ERROR_FAIL;
+	}
```

Comment:
As noted above, please replace this by an assert. It is an invariant.

```suggestion
	/* The register is cacheable, so the cache entry must be valid now. */
	assert(reg->is_valid);

```

Reply:
Addressed.